x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java

```diff
@@ -90,12 +90,16 @@ public interface Role {
      * @param checkForIndexPatterns check permission grants for the set of index patterns
      * @param allowRestrictedIndices if {@code true} then checks permission grants even for restricted indices by index matching
      * @param checkForPrivileges check permission grants for the set of index privileges
-     * @return an instance of {@link ResourcePrivilegesMap}
+     * @param resourcePrivilegesMapBuilder out-parameter for returning the details on which privilege over which resource is granted or not.
+     *                                     Can be {@code null} when no such details are needed so the method can return early, after
+     *                                     encountering the first privilege that is not granted over some resource.
+     * @return {@code true} when all the privileges are granted over all the resources, or {@code false} otherwise
      */
-    ResourcePrivilegesMap checkIndicesPrivileges(
+    boolean checkIndicesPrivileges(
         Set<String> checkForIndexPatterns,
         boolean allowRestrictedIndices,
-        Set<String> checkForPrivileges
+        Set<String> checkForPrivileges,
+        @Nullable ResourcePrivilegesMap.Builder resourcePrivilegesMapBuilder
     );
```

Comment:
I would like to take some space to explain my interface decisions for: `Role#checkIndicesPrivilegeges`, `Role#checkApplicationResourcePrivileges`, `IndicesPermission#checkREsourcePrivileges`, and `ApplicationPermission#checkResourcePrivileges`.

These methods must be aware of the mode of operation, ie whether to short-circuit or not, because they must return early in the short-circuit mode (in addition, the response "type" is different between the 2 modes) . So, one of the parameters of these methods must express this operation mode.

In this case, I think there are two implementation options to go about it:
* pass a dedicated parameter, to signal the short-circuit mode, and make the return value contain an optional field which is `null` when in the short-circuit mode; eg `RBACEngine#checkPrivileges(..., boolean runDetailedCheck, ..., ActionListener<PrivilegesCheckResult> listener)` where `PrivilegesCheckResult` has two fields: `boolean allChecksSuccess` and an optional `Details details`.
* return 2 values, rather than a single one with an optional `null` field. The first return value is the `allChecksSuccess` flag. The second returned value is via an out-parameter 👻 . When the out-parameter is non-`null`, it is used to return the privileges details, ie it works in the regular, non-short-circuit mode; otherwise, if `null`, no check details are returned and the operation mode is short-circuit. 

I had 2 reasons to choose the second option for these methods (unlike `RBACEngine#checkPrivileges`):
* I disliked having a `boolean` parameter going from the REST handler all the layers through to the permission level inside a role.
* `ResourcePrivilegesMap` has a boolean field `allowAll`. Its value is currently driven by the individual privilege check results. But In the short-circuit mode, it becomes separated from the check results, as it can be `true`/`false` when the individual privilege check results are `null`. With such a separated `allowAll` parameter, keeping it still a member of the `ResourcePrivilegesMap` felt janky (what does it mean to update its value in the non-short-circuit mode, for every privilege result, when its value can technically also be set externally - the encapsulation was gone).

I hope the scare of an out-parameter is not too great.
I think this is the more elegant implementation of the two. I like that these `check*` methods always return a simple boolean. Also, the out parameter works OK in the application check section in `RBACEngine#checkPrivileges` and in the `LimitedRole` class, I think.

Reply:
I noticed the out parameters on my first reading of the PR. I would probably approach it differently. But what is proposed in this PR works as well. Plus you already considered the other option and ruled it out with reasons. Also your reasoning about the standalone `allowAll` field and the individual check results of `ResourcePrivilegesMap` is a good one. 

In summary, I am OK with the out parameters. 

PS: In future, we may need support caching for LimitedRole (now it is only supported for SimpleRole). For that I am thinking that the result intersecting between `role` and `limited-by-role` should be done at the overall `PrivilegesCheckResult` level rather than the individual checks level (and also move most `RBCEngine#checkPrivileges` to `Role` level). This has nothing to do with the current PR. Just thinking out loud in case it is something you also thought about.

Reply:
> In summary, I am OK with the out parameters.

🙇 

> PS: In future, we may need support caching for LimitedRole (now it is only supported for SimpleRole). For that I am thinking that the result intersecting between role and limited-by-role should be done at the overall PrivilegesCheckResult level rather than the individual checks level (and also move most RBCEngine#checkPrivileges to Role level). This has nothing to do with the current PR. Just thinking out loud in case it is something you also thought about.

No I haven't considered something like that. I'm not sure I follow either. My first approach would be to investigate whether the `LimitedRole` can be cached like the `Role` can.